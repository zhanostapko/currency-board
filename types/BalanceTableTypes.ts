export type BalanceItem = {
  id: string;
  currency: string;
  amount: string;
};

export type BalancesTableProps = {
  data: BalanceItem[];
  columnCount: number;
  showDeleteModal: (currency: string) => void;
};
