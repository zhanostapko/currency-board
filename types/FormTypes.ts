export type RegisterActionType = {
  error: string | null;
  success: boolean;
  email: string;
};

export type LoginActionType = {
  error: string | null;
  step: "login" | "otp";
  email: string;
  otp: number | null;
};

export type OtpActionType = {
  error: string | null;
  step: "otp";
};
