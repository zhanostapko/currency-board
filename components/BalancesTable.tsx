import React from "react";
import Image from "next/image";
import { BalanceItem } from "@/types/BalanceTableTypes";

type BalancesTableProps = {
  data: BalanceItem[];
  columnCount: number;
  showDeleteModal: (currency: string) => void;
  error?: string | null;
};

const BalancesTable = ({
  data,
  columnCount,
  showDeleteModal,
  error,
}: BalancesTableProps) => {
  const groupedData = data.reduce<BalanceItem[][]>((acc, item, index) => {
    const rowIndex = Math.floor(index / columnCount);
    if (!acc[rowIndex]) acc[rowIndex] = [];
    acc[rowIndex].push(item);
    return acc;
  }, []);

  return (
    <>
      {" "}
      {error ? (
        <div className="w-full border border-red-500 bg-red-50 text-red-700 text-center font-medium p-4 rounded-md">
          ‚ùå Failed to fetch data. Please try again later.
        </div>
      ) : (
        <div className="w-full overflow-x-auto">
          <table className="w-full border-collapse rounded-lg shadow-sm table-fixed">
            <thead>
              <tr className="bg-gray-200 text-gray-900 text-lg font-semibold">
                {Array.from({ length: columnCount }, (_, i) => (
                  <React.Fragment key={i}>
                    <th className="px-6 py-4 text-left">Currency</th>
                    <th className="px-6 py-4 text-right">Balance</th>
                  </React.Fragment>
                ))}
              </tr>
            </thead>
            <tbody>
              {groupedData.length > 0 ? (
                groupedData.map((row, rowIndex) => (
                  <tr
                    key={rowIndex}
                    className="even:bg-gray-100 hover:bg-gray-200 transition"
                  >
                    {row.map(({ currency, amount }) => (
                      <React.Fragment key={currency}>
                        <td className="px-6 py-4 text-gray-700 font-medium text-left">
                          {currency}
                        </td>
                        <td className="px-6 py-4 text-gray-700 font-medium flex justify-end items-center gap-2">
                          {amount}
                          <button
                            onClick={() => showDeleteModal(currency)}
                            className="text-red-500 hover:text-red-700 transition"
                          >
                            <Image
                              src="trashIcon.svg"
                              alt="Delete"
                              width="14"
                              height="14"
                            />
                          </button>
                        </td>
                      </React.Fragment>
                    ))}
                  </tr>
                ))
              ) : (
                <tr>
                  <td
                    colSpan={columnCount * 2}
                    className="px-6 py-4 text-center text-gray-500"
                  >
                    üîé No balances found for your search.
                  </td>
                </tr>
              )}
            </tbody>
          </table>
        </div>
      )}
    </>
  );
};

export default BalancesTable;
