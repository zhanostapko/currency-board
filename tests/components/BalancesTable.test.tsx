import { render, screen, fireEvent } from "@testing-library/react";
import BalancesTable from "../../components/BalancesTable";
import { BalanceItem } from "../../types/BalanceTableTypes";

describe("BalancesTable Component", () => {
  const mockData: BalanceItem[] = [
    { id: "1", currency: "USD", amount: "100" },
    { id: "2", currency: "EUR", amount: "200" },
  ];
  const mockShowDeleteModal = jest.fn();

  test("renders table with data", () => {
    const { asFragment } = render(
      <BalancesTable
        data={mockData}
        columnCount={2}
        showDeleteModal={mockShowDeleteModal}
      />
    );

    expect(screen.getByText("USD")).toBeInTheDocument();
    expect(screen.getByText("100")).toBeInTheDocument();
    expect(screen.getByText("EUR")).toBeInTheDocument();
    expect(screen.getByText("200")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test("shows error message when error prop is passed", () => {
    const { asFragment } = render(
      <BalancesTable
        data={mockData}
        columnCount={2}
        showDeleteModal={mockShowDeleteModal}
        error="Test error"
      />
    );

    expect(
      screen.getByText("❌ Failed to fetch data. Please try again later.")
    ).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test("calls showDeleteModal on delete button click", () => {
    const { asFragment } = render(
      <BalancesTable
        data={mockData}
        columnCount={2}
        showDeleteModal={mockShowDeleteModal}
      />
    );

    fireEvent.click(screen.getAllByRole("button")[0]);
    expect(mockShowDeleteModal).toHaveBeenCalledWith("USD");
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders empty state when no data is provided", () => {
    const { asFragment } = render(
      <BalancesTable
        data={[]}
        columnCount={2}
        showDeleteModal={mockShowDeleteModal}
      />
    );

    expect(
      screen.getByText("🔎 No balances found for your search.")
    ).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
