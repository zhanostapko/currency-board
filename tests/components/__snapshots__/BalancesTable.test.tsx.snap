// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BalancesTable Component calls showDeleteModal on delete button click 1`] = `
<DocumentFragment>
   
  <div
    class="w-full overflow-x-auto"
  >
    <table
      class="w-full border-collapse rounded-lg shadow-sm table-fixed"
    >
      <thead>
        <tr
          class="bg-gray-200 text-gray-900 text-lg font-semibold"
        >
          <th
            class="px-6 py-4 text-left"
          >
            Currency
          </th>
          <th
            class="px-6 py-4 text-right"
          >
            Balance
          </th>
          <th
            class="px-6 py-4 text-left"
          >
            Currency
          </th>
          <th
            class="px-6 py-4 text-right"
          >
            Balance
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="even:bg-gray-100 hover:bg-gray-200 transition"
        >
          <td
            class="px-6 py-4 text-gray-700 font-medium text-left"
          >
            USD
          </td>
          <td
            class="px-6 py-4 text-gray-700 font-medium flex justify-end items-center gap-2"
          >
            100
            <button
              class="text-red-500 hover:text-red-700 transition"
            >
              <img
                alt="Delete"
                data-nimg="1"
                decoding="async"
                height="14"
                loading="lazy"
                src="trashIcon.svg"
                style="color: transparent;"
                width="14"
              />
            </button>
          </td>
          <td
            class="px-6 py-4 text-gray-700 font-medium text-left"
          >
            EUR
          </td>
          <td
            class="px-6 py-4 text-gray-700 font-medium flex justify-end items-center gap-2"
          >
            200
            <button
              class="text-red-500 hover:text-red-700 transition"
            >
              <img
                alt="Delete"
                data-nimg="1"
                decoding="async"
                height="14"
                loading="lazy"
                src="trashIcon.svg"
                style="color: transparent;"
                width="14"
              />
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`BalancesTable Component renders empty state when no data is provided 1`] = `
<DocumentFragment>
   
  <div
    class="w-full overflow-x-auto"
  >
    <table
      class="w-full border-collapse rounded-lg shadow-sm table-fixed"
    >
      <thead>
        <tr
          class="bg-gray-200 text-gray-900 text-lg font-semibold"
        >
          <th
            class="px-6 py-4 text-left"
          >
            Currency
          </th>
          <th
            class="px-6 py-4 text-right"
          >
            Balance
          </th>
          <th
            class="px-6 py-4 text-left"
          >
            Currency
          </th>
          <th
            class="px-6 py-4 text-right"
          >
            Balance
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="px-6 py-4 text-center text-gray-500"
            colspan="4"
          >
            üîé No balances found for your search.
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`BalancesTable Component renders table with data 1`] = `
<DocumentFragment>
   
  <div
    class="w-full overflow-x-auto"
  >
    <table
      class="w-full border-collapse rounded-lg shadow-sm table-fixed"
    >
      <thead>
        <tr
          class="bg-gray-200 text-gray-900 text-lg font-semibold"
        >
          <th
            class="px-6 py-4 text-left"
          >
            Currency
          </th>
          <th
            class="px-6 py-4 text-right"
          >
            Balance
          </th>
          <th
            class="px-6 py-4 text-left"
          >
            Currency
          </th>
          <th
            class="px-6 py-4 text-right"
          >
            Balance
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="even:bg-gray-100 hover:bg-gray-200 transition"
        >
          <td
            class="px-6 py-4 text-gray-700 font-medium text-left"
          >
            USD
          </td>
          <td
            class="px-6 py-4 text-gray-700 font-medium flex justify-end items-center gap-2"
          >
            100
            <button
              class="text-red-500 hover:text-red-700 transition"
            >
              <img
                alt="Delete"
                data-nimg="1"
                decoding="async"
                height="14"
                loading="lazy"
                src="trashIcon.svg"
                style="color: transparent;"
                width="14"
              />
            </button>
          </td>
          <td
            class="px-6 py-4 text-gray-700 font-medium text-left"
          >
            EUR
          </td>
          <td
            class="px-6 py-4 text-gray-700 font-medium flex justify-end items-center gap-2"
          >
            200
            <button
              class="text-red-500 hover:text-red-700 transition"
            >
              <img
                alt="Delete"
                data-nimg="1"
                decoding="async"
                height="14"
                loading="lazy"
                src="trashIcon.svg"
                style="color: transparent;"
                width="14"
              />
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`BalancesTable Component shows error message when error prop is passed 1`] = `
<DocumentFragment>
   
  <div
    class="w-full border border-red-500 bg-red-50 text-red-700 text-center font-medium p-4 rounded-md"
  >
    ‚ùå Failed to fetch data. Please try again later.
  </div>
</DocumentFragment>
`;
